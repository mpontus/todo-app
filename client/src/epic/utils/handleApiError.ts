import { OperatorFunction, of } from "rxjs";
import { Action } from "redux";
import { catchError } from "rxjs/operators";
import { RequestError } from "../../model/RequestError";

/**
 * Helper function to handle API error
 *
 * Wraps error caught from source observable into RequestError and
 * passes it onto specified action creator before merging it into the
 * result stream.
 */
export const handleApiError = <T, R>(
  failureActionCreator: (e: any) => R
): OperatorFunction<T, T | R> =>
  catchError(error =>
    of(failureActionCreator(RequestError.fromApiError(error)))
  );
