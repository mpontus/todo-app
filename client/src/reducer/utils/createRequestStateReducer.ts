import { Reducer } from "redux";
import { Action } from "../../action";

interface RequestState<ErrorType> {
  loading: boolean;
  success: boolean;
  error?: ErrorType;
}

const initialState = {
  loading: false,
  success: false
};

export const createRequestStateReducer = <
  BaseAction extends Action,
  RequestAction extends BaseAction,
  SuccessAction extends BaseAction,
  FailureAction extends BaseAction,
  ErrorType extends Error
>(
  requestActionPredicate: (value: BaseAction) => value is RequestAction,
  successActionPredicate: (value: BaseAction) => value is SuccessAction,
  failureActionPredicate: (value: BaseAction) => value is FailureAction,
  errorSelector: (value: FailureAction) => ErrorType
): Reducer<RequestState<ErrorType>, BaseAction> => (
  state = initialState,
  action
) => {
  if (requestActionPredicate(action)) {
    return {
      loading: true,
      success: false
    };
  }

  if (successActionPredicate(action)) {
    return {
      loading: false,
      success: true
    };
  }

  if (failureActionPredicate(action)) {
    return {
      loading: false,
      success: false,
      error: errorSelector(action)
    };
  }

  return state;
};
