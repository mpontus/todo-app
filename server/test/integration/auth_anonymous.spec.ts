import supertest from 'supertest';
import { initApp } from '../utils/initApp';
import { resetDb } from '../utils/resetDb';

let expressApp: any;
let nestApp: any;

beforeAll(async () => {
  ({ expressApp, nestApp } = await initApp());
});

afterAll(() => nestApp.close());

beforeEach(resetDb);

describe('anonymous authentication', () => {
  it('should be successful', async () => {
    const response = await supertest(expressApp)
      .post('/auth/anonymous')
      .expect(201);

    expect(response.body).toMatchSnapshot({
      token: expect.any(String),
    });
  });
});
