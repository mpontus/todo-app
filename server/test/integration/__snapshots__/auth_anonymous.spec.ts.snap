// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`anonymous authentication should be successful 1`] = `
Object {
  "token": Any<String>,
}
`;
